patch was not used message does not adequately describe reason

Heya, not _sure_ whether we consider usability issues bugs, but, it seemed the closest option.

When configuring `[patch]`es there are a variety of reasons the patch may not be used, resulting in a message like the following:

```
warning: Patch `SOMETHING` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
```

This gives a list of reasons that can be particularly difficult to debug, as well as the recommendation that `cargo update` may help, which is particularly helpful when this message is printed by `cargo update`. As an alternative, it would be great to print the _specific_ reason the `[patch]` wasn't used, and a proposed resolution, which I believe is in-line with other compiler error improvements.

Some possible examples:
```
warning: Patch `SOMETHING` was not used in the crate graph.
Patch version "0.14.0" from `../SOMETHING` does not meet package constraint ">0.15.0",
this may be resolved by updating the `SOMETHING` package version

warning: Patch `SOMETHING` was not used in the crate graph.
Patch version "0.14.0" does not match the package version in Cargo.lock,
this may be resolved by running `cargo update` to update the cargo lock file.
```

Related to: https://github.com/rust-lang/cargo/issues/8400

Transferred.

@ryankurte it would be helpful if you could put together an example that doesn't print a helpful error message. Please also include which version of Cargo you are using. The messages were improved recently (#8248), and the example where the patch does not satisfy the dependency requirement should say something like this:

```
[ERROR] failed to resolve patches for `https://github.com/rust-lang/crates.io-index`
Caused by:
  patch for `bar` in `https://github.com/rust-lang/crates.io-index` failed to resolve
Caused by:
  The patch location `[..]/foo/bar` contains a `bar` package with version `0.1.0`, 
  but the patch definition requires `^0.1.1`.
  Check that the version in the patch location is what you expect, and update the 
  patch definition to match.
```

--- a/src/cargo/core/registry.rs
+++ b/src/cargo/core/registry.rs
@@ -413,11 +413,12 @@ impl<'cfg> PackageRegistry<'cfg> {
         self.patches_locked = true;
     }
 
-    pub fn patches(&self) -> Vec<Summary> {
-        self.patches
-            .values()
-            .flat_map(|v| v.iter().cloned())
-            .collect()
+    /// Gets all patches grouped by the source URLS they are going to patch.
+    ///
+    /// These patches are mainly collected from [`patch`](Self::patch).
+    /// They might not be the same as patches actually used during dependency resolving.
+    pub fn patches(&self) -> &HashMap<CanonicalUrl, Vec<Summary>> {
+        &self.patches
     }
 
     fn load(&mut self, source_id: SourceId, kind: Kind) -> CargoResult<()> {

--- a/src/cargo/ops/resolve.rs
+++ b/src/cargo/ops/resolve.rs
@@ -29,7 +29,7 @@ use crate::util::errors::CargoResult;
 use crate::util::{profile, CanonicalUrl};
 use anyhow::Context as _;
 use log::{debug, trace};
-use std::collections::HashSet;
+use std::collections::{HashMap, HashSet};
 
 /// Result for `resolve_ws_with_opts`.
 pub struct WorkspaceResolve<'cfg> {
@@ -467,25 +467,17 @@ pub fn resolve_with_previous<'cfg>(
             .require(Feature::public_dependency())
             .is_ok(),
     )?;
-    resolved.register_used_patches(&registry.patches());
-    if register_patches {
-        // It would be good if this warning was more targeted and helpful
-        // (such as showing close candidates that failed to match). However,
-        // that's not terribly easy to do, so just show a general help
-        // message.
-        let warnings: Vec<String> = resolved
-            .unused_patches()
-            .iter()
-            .map(|pkgid| format!("Patch `{}` was not used in the crate graph.", pkgid))
-            .collect();
-        if !warnings.is_empty() {
-            ws.config().shell().warn(format!(
-                "{}\n{}",
-                warnings.join("\n"),
-                UNUSED_PATCH_WARNING
-            ))?;
-        }
+    let patches: Vec<_> = registry
+        .patches()
+        .values()
+        .flat_map(|v| v.iter().cloned())
+        .collect();
+    resolved.register_used_patches(&patches[..]);
+
+    if register_patches && !resolved.unused_patches().is_empty() {
+        emit_warnings_of_unused_patches(ws, &resolved, &registry)?;
     }
+
     if let Some(previous) = previous {
         resolved.merge_from(previous)?;
     }
@@ -756,4 +748,80 @@ fn master_branch_git_source(id: PackageId, resolve: &Resolve) -> Option<PackageI
         }
     }
     None
+}
+
+/// Emits warnings of unused patches case by case.
+///
+/// This function does its best to provide more targeted and helpful
+/// (such as showing close candidates that failed to match). However, that's
+/// not terribly easy to do, so just show a general help message if we cannot.
+fn emit_warnings_of_unused_patches(
+    ws: &Workspace<'_>,
+    resolve: &Resolve,
+    registry: &PackageRegistry<'_>,
+) -> CargoResult<()> {
+    const MESSAGE: &str = "was not used in the crate graph.";
+
+    // Patch package with the source URLs being patch
+    let mut patch_pkgid_to_urls = HashMap::new();
+    for (url, summaries) in registry.patches().iter() {
+        for summary in summaries.iter() {
+            patch_pkgid_to_urls
+                .entry(summary.package_id())
+                .or_insert_with(HashSet::new)
+                .insert(url);
+        }
+    }
+
+    // pkg name -> all source IDs of under the same pkg name
+    let mut source_ids_grouped_by_pkg_name = HashMap::new();
+    for pkgid in resolve.iter() {
+        source_ids_grouped_by_pkg_name
+            .entry(pkgid.name())
+            .or_insert_with(HashSet::new)
+            .insert(pkgid.source_id());
+    }
+
+    let mut unemitted_unused_patches = Vec::new();
+    for unused in resolve.unused_patches().iter() {
+        // Show alternative source URLs if the source URLs being patch
+        // cannot not be found in the crate graph.
+        match (
+            source_ids_grouped_by_pkg_name.get(&unused.name()),
+            patch_pkgid_to_urls.get(unused),
+        ) {
+            (Some(ids), Some(patched_urls))
+                if ids
+                    .iter()
+                    .all(|id| !patched_urls.contains(id.canonical_url())) =>
+            {
+                use std::fmt::Write;
+                let mut msg = String::new();
+                writeln!(&mut msg, "Patch `{}` {}", unused, MESSAGE)?;
+                write!(
+                    &mut msg,
+                    "Perhaps you misspell the source URL being patched.\n\
+                    Possible URLs for `[patch.<URL>]`:",
+                )?;
+                for id in ids.iter() {
+                    write!(&mut msg, "\n    {}", id.display_registry_name())?;
+                }
+                ws.config().shell().warn(msg)?;
+            }
+            _ => unemitted_unused_patches.push(unused),
+        }
+    }
+
+    // Show general help message.
+    if !unemitted_unused_patches.is_empty() {
+        let warnings: Vec<_> = unemitted_unused_patches
+            .iter()
+            .map(|pkgid| format!("Patch `{}` {}", pkgid, MESSAGE))
+            .collect();
+        ws.config()
+            .shell()
+            .warn(format!("{}\n{}", warnings.join("\n"), UNUSED_PATCH_WARNING))?;
+    }
+
+    return Ok(());
 }

--- a/tests/testsuite/patch.rs
+++ b/tests/testsuite/patch.rs
@@ -358,10 +358,10 @@ fn unused() {
             "\
 [UPDATING] `dummy-registry` index
 [WARNING] Patch `bar v0.2.0 ([CWD]/bar)` was not used in the crate graph.
-[..]
-[..]
-[..]
-[..]
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
 [DOWNLOADING] crates ...
 [DOWNLOADED] bar v0.1.0 [..]
 [COMPILING] bar v0.1.0
@@ -374,10 +374,10 @@ fn unused() {
         .with_stderr(
             "\
 [WARNING] Patch `bar v0.2.0 ([CWD]/bar)` was not used in the crate graph.
-[..]
-[..]
-[..]
-[..]
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
 [FINISHED] [..]
 ",
         )
@@ -392,6 +392,60 @@ fn unused() {
         toml["patch"]["unused"][0]["version"].as_str(),
         Some("0.2.0")
     );
+}
+
+#[cargo_test]
+fn unused_with_mismatch_source_being_patched() {
+    registry::alt_init();
+    Package::new("bar", "0.1.0").publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+                authors = []
+
+                [dependencies]
+                bar = "0.1.0"
+
+                [patch.alternative]
+                bar = { path = "bar" }
+
+                [patch.crates-io]
+                bar = { path = "baz" }
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .file("bar/Cargo.toml", &basic_manifest("bar", "0.2.0"))
+        .file("bar/src/lib.rs", "not rust code")
+        .file("baz/Cargo.toml", &basic_manifest("bar", "0.3.0"))
+        .file("baz/src/lib.rs", "not rust code")
+        .build();
+
+    p.cargo("build")
+        .with_stderr(
+            "\
+[UPDATING] `dummy-registry` index
+[WARNING] Patch `bar v0.2.0 ([CWD]/bar)` was not used in the crate graph.
+Perhaps you misspell the source URL being patched.
+Possible URLs for `[patch.<URL>]`:
+    crates-io
+[WARNING] Patch `bar v0.3.0 ([CWD]/baz)` was not used in the crate graph.
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
+[DOWNLOADING] crates ...
+[DOWNLOADED] bar v0.1.0 [..]
+[COMPILING] bar v0.1.0
+[COMPILING] foo v0.0.1 ([CWD])
+[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]
+",
+        )
+        .run();
 }
 
 #[cargo_test]
@@ -477,10 +531,10 @@ fn unused_from_config() {
             "\
 [UPDATING] `dummy-registry` index
 [WARNING] Patch `bar v0.2.0 ([CWD]/bar)` was not used in the crate graph.
-[..]
-[..]
-[..]
-[..]
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
 [DOWNLOADING] crates ...
 [DOWNLOADED] bar v0.1.0 [..]
 [COMPILING] bar v0.1.0
@@ -493,10 +547,10 @@ fn unused_from_config() {
         .with_stderr(
             "\
 [WARNING] Patch `bar v0.2.0 ([CWD]/bar)` was not used in the crate graph.
-[..]
-[..]
-[..]
-[..]
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
 [FINISHED] [..]
 ",
         )
@@ -550,10 +604,10 @@ fn unused_git() {
 [UPDATING] git repository `file://[..]`
 [UPDATING] `dummy-registry` index
 [WARNING] Patch `bar v0.2.0 ([..])` was not used in the crate graph.
-[..]
-[..]
-[..]
-[..]
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
 [DOWNLOADING] crates ...
 [DOWNLOADED] bar v0.1.0 [..]
 [COMPILING] bar v0.1.0
@@ -566,10 +620,10 @@ fn unused_git() {
         .with_stderr(
             "\
 [WARNING] Patch `bar v0.2.0 ([..])` was not used in the crate graph.
-[..]
-[..]
-[..]
-[..]
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
 [FINISHED] [..]
 ",
         )
@@ -752,10 +806,10 @@ fn add_ignored_patch() {
         .with_stderr(
             "\
 [WARNING] Patch `bar v0.1.1 ([CWD]/bar)` was not used in the crate graph.
-[..]
-[..]
-[..]
-[..]
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
 [FINISHED] dev [unoptimized + debuginfo] target(s) in [..]",
         )
         .run();
@@ -763,10 +817,10 @@ fn add_ignored_patch() {
         .with_stderr(
             "\
 [WARNING] Patch `bar v0.1.1 ([CWD]/bar)` was not used in the crate graph.
-[..]
-[..]
-[..]
-[..]
+Check that [..]
+with the [..]
+what is [..]
+version. [..]
 [FINISHED] [..]",
         )
         .run();
@@ -1714,10 +1768,9 @@ fn two_semver_compatible() {
         .with_stderr(
             "\
 warning: Patch `bar v0.1.1 [..]` was not used in the crate graph.
-Check that [..]
-with the [..]
-what is [..]
-version. [..]
+Perhaps you misspell the source URL being patched.
+Possible URLs for `[patch.<URL>]`:
+    [CWD]/bar
 [FINISHED] [..]",
         )
         .run();
@@ -1769,10 +1822,9 @@ fn multipatch_select_big() {
         .with_stderr(
             "\
 warning: Patch `bar v0.1.0 [..]` was not used in the crate graph.
-Check that [..]
-with the [..]
-what is [..]
-version. [..]
+Perhaps you misspell the source URL being patched.
+Possible URLs for `[patch.<URL>]`:
+    [CWD]/bar
 [FINISHED] [..]",
         )
         .run();

